510 scheduler

PsiOmegaDelta opened the pull request:
<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->

Ports vg's 510 scheduler tweaks. Unsure if it's functioning as awesomely as on other 510 utilizing servers but my laptop could handle Nar-Sie, a maximum level singularity, and a supermatter cascade with some grace (despite the hundreds of lighting runtimes caused by said cascade).

Most if not all of the above likely need some performance tweaking (and bug fixing), with or without this scheduler. At least running a couple of meteor showers in parallel didn't seem to bother my laptop much.

All in all, things don't seem to be worse.

Mloc commented:
Might testrun this on live to see how it goes. Can't really judge just by the code, but it's probably fine.

PJB3005 commented:
Our scheduler isn't that up to date compared to somewhere like paradise, where @Krausus did a lot of work on the scheduler. Might wanna consider looking into it.

PsiOmegaDelta commented:
It is sufficiently up-to-date to be 510 utilizing at least. Unsure when I'll have a big enough block of time to port the Paradise scheduler.

Fox-McCloud commented:
@Tastyfish also did some updates on the scheduler to further improve it and make it compatible with 510.

PsiOmegaDelta committed:
Ports Paradise's 510 scheduler.

PsiOmegaDelta commented:
Very well. Have now switched to the Paradise scheduler and removed all traces of btime.
The server no longer has to be booted with the --trusted flag.

mwerezak commented on code/controllers/ProcessScheduler/core/process.dm:
```diff
@@ -208,17 +199,12 @@
 	if (main.getCurrentTickElapsedTime() > main.timeAllowance)
```
Does `getCurrentTickElapsedTime()` have a purpose anymore now that we have `TimeOfTick`? It seems like it tries to do what `TimeOfTick` does only more complicated.

mwerezak replied:
Nevermind, missed the edit to that proc below.

mwerezak commented on code/controllers/ProcessScheduler/core/processScheduler.dm:
```diff
@@ -349,29 +343,23 @@ var/global/datum/controller/processScheduler/processScheduler
 		return 0
 	else
-		return TimeOfHour - currentTickStart
+		return TimeOfTick
```
I was going to comment that since `world.tick_usage` resets back to 0 at the beginning of each new tick, this branching logic doesn't seem necessary, but then I noticed that it also calls `updateCurrentTickData()`. That proc name seems kind of inappropriate given this side effect, but I guess that's not on topic for this PR.

EDIT: Then I noticed that `updateCurrentTickData()` is called from ` /datum/controller/processScheduler/process()` anyways. Is it really needed here? It would be nice if we could just replace this proc with `TimeOfTick`.

PsiOmegaDelta commented:
¯\\_(ツ)_/¯

mwerezak commented:
:+1:

Techhead0 commented on code/controllers/ProcessScheduler/core/processScheduler.dm:
```diff
@@ -349,29 +343,23 @@ var/global/datum/controller/processScheduler/processScheduler
 	// Time allowance goes down linearly with world.cpu.
 	var/tmp/error = cpuAverage - 100
-	var/tmp/timeAllowanceDelta = sign(error) * -0.5 * world.tick_lag * max(0, 0.001 * abs(error))
+	var/tmp/timeAllowanceDelta = SIMPLE_SIGN(error) * -0.5 * world.tick_lag * max(0, 0.001 * abs(error))
```
Why isn't this expression just `-0.005 * world.tick_lag * error`?

Techhead0 commented on code/__defines/process_scheduler.dm:
```diff
@@ -11,10 +11,9 @@
 // This references src directly to work around a weird bug with try/catch
 #define SCHECK_EVERY(this_many_calls) if(++src.calls_since_last_scheck >= this_many_calls) sleepCheck()
-#define SCHECK SCHECK_EVERY(50)
+#define SCHECK sleepCheck()
```
`sleepCheck()` has a comment stating `// Do not call this directly - use SHECK or SCHECK_EVERY`. Besides the misspelling, the comment might need updating.

PsiOmegaDelta replied:
There are no 'direct' callers of sleepCheck(), everything still use the SCHECK macro.

mwerezak commented:
@Mloc would be nice if you could review this as well.
